astroid/nodes/scoped_nodes/scoped_nodes.py

```diff
@@ -1979,6 +1979,19 @@ def frame(self: T, *, future: Literal[None, True] = None) -> T:
         """
         return self
 
+    @decorators_mod.cachedproperty
+    def tolineno(self) -> Optional[int]:
+        """The last line that this node appears on in the source code."""
+        if self.end_lineno is not None:
+            return self.end_lineno
+        last_child = self.last_child()
+        # This is a hack to approximate the final line of a node
+        # with only a docstring, as those are not represented as child.
+        if isinstance(last_child, Arguments):
```

Comment:
I do wonder how resilient that will be. `last_child` can be any astroid node. For `FunctionDef` those fields are `("decorators", "bases", "keywords", "body")`.

Reply:
https://github.com/PyCQA/astroid/blob/56f5f055e4847e9fc2b74162ccad8f2a11db3bfc/astroid/nodes/node_ng.py#L445-L457

This is copied from the original implementation. However, since `FunctionDef` always has an `Arguments` `last_child is None` doesn't work. This tries to replicate that.

Reply:
What if the function has a return type annotation?
```py
def func() -> int:
    """Hello World"""
```
`if isinstance(last_child, Arguments)` won't work then and it would default to `last_child.tolineno`.

Reply:
Hmm, I guess this is correct.
```python
def func() -> None:
    """Hello World"""
```

But I mean how often do people actually write function and classes with only a `docstring`. This problem existed in previous versions as well, as `tolineno` was broken for those versions as well. Only because we are now using `end_lineno` we noticed this. I wonder how big of a problem this actually is.

That said, I'm also fine with putting more effort into #1276 as finding all edge cases here will probably take considerable time. We will need to take either approach before we can accept `2.1x` in `pylint` though.

Reply:
This might be relative strait forward to solve (at least compared to the issue with ClassDef` below).

We know `last_child` will at least be `Arguments`, so we don't need to care about decorators anymore. Something like this could work
```py
if not body and last_child:
    ...
```

Reply:
```python
def func() -> None:
    """Hello World"""
```

I don't think that will always work. `last_child` is a `Const` `None` for this example. We could some comparison with `last_child.tolineno` and the `end_lineno` of the `FunctionDef` and see if they are similar. But then when the return type annotation is on a new line we would still wrongly infer the line number.
I think this is getting unnecessarily complicated knowing that #1276 is on the horizon. I think it makes more sense to help land that quickly and then revisit this.

In the meantime we could revert the commit that changed the `tolineno` behaviour to make `astroid` usable for `pylint`, but that might also be a little too much.